### PR TITLE
[JUJU-2024] avoid microk8s permission check when bootstrapping lxd

### PR DESCRIPTION
**Pull request by basebandit**
When bootstrapping an LXD controller and the current `$USER` is not a member of the microk8s group, the bootstrap  fails due to this. This PR adds a check for non-microk8s cloud in the `k8s` provider's  `CloudDetector` implementation (which is responsible for checking for microk8s permission)  to avoid returning an error if the cloud we are bootstrapping is a non-microk8s cloud, i.e when boostrapping an LXD controller. 
NOTE: This error is intermittent, it might take you a couple of tries before you reproduce it. 

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- ~[ ] Comments saying why design decisions were made~
- ~[ ] Go unit tests, with comments saying what you're testing~
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

```sh
sudo deluser $USER microk8s
su - $USER
juju bootstrap localhost
```
You should be able to bootstrap a lxd controller, even if you are not a microk8s group member.
All relevant unit tests should pass

## Bug reference

https://bugs.launchpad.net/juju/+bug/1937985

### CODE REVIEW

**Comment by barrettj12** [comment]
This doesn't quite seem like the right fix to me. The whole point of these provider interfaces (and interfaces in general) is to avoid the need for messy case analysis.

If we have specified LXD at bootstrap time, then it should not be invoking the "kubernetes provider's cloud detection logic" at all. We should be invoking the LXD provider's cloud detection logic instead. My question would be why we are even talking to the k8s provider **at all** when trying to bootstrap on LXD.

In other words, the `cloudDetector` here should be the LXD one
```go
		cloudDetector, ok := bootstrapFuncs.CloudDetector(provider)
		if !ok {
			continue
		}
		cloud, err := cloudDetector.DetectCloud(c.Cloud)
```
If we're somehow getting the k8s cloud detector instead, then we should work out why and fix the root cause.

**Comment by wallyworld** [comment]
The correct fix is to make the microk8s detection return false when the user group issue is present and not an error.

**Comment by basebandit** [comment]
Thank you for the feedback. This was a bandaid fix to gather feedback. I am actively looking into Ian's suggestion and making sure the experience for microk8s user remains as was.

**Comment by basebandit** [comment]
I have find an alternative fix instead of changing the signature of the the `DetectCloud` to return false, In the microk8s provider's implementation of CloudDetector's DetectCloud behaviour, we add a check for non-microk8s clouds and avoid returning an error for such clouds. Also this error is intermittent in nature because of the fact that the providers are stored in a map and when rangin over a map the order is not always guaranteed.
```

**Comment by basebandit** [comment]
The errors need to be shown still for when you are bootstrapping a microk8s cloud and something is not right. Still thinking how to implement that now that we are no longer returning the error. Probably could introduce a context.Stdblah in the provider to write the errors to stderr??

**Comment by wallyworld** [comment]
I was looking at the code in between meetings and misread the API so apologies for the misinformation. The bool return was for getting the cloud detector off the provider, not the detection operation itself.

The `DetectCloud()` APi needs to return `NotFound` if the named cloud is not possible. See the bootstrap code below skips over clouds with not found.

```
		cloudDetector, ok := bootstrapFuncs.CloudDetector(provider)
		if !ok {
			continue
		}
		cloud, err := cloudDetector.DetectCloud(c.Cloud)
		if errors.IsNotFound(err) {
			continue
```
Looking at the k8s DetectCloud:

```
	mk8sCloud, err := p.builtinCloudGetter(p.cmdRunner)
	if err == nil && name == k8s.K8sCloudMicrok8s {
		return mk8sCloud, nil
	}
	if !errors.IsNotFound(err) && err != nil {
		return cloud.Cloud{}, errors.Trace(err)
	}
```

We just need to tweak the logic so that if the cloud is microk8s and we get the user group error, we return not found so that the bootstrap processing skips the cloud.

File: caas/kubernetes/provider/detectcloud.go [hunk 1]
```diff
@@ -54,7 +54,13 @@ func (p kubernetesEnvironProvider) DetectCloud(name string) (cloud.Cloud, error)
 	if err == nil && name == k8s.K8sCloudMicrok8s {
 		return mk8sCloud, nil
 	}
+
 	if !errors.IsNotFound(err) && err != nil {
+		// if the cloud is not microk8s and we get the user group error, return not found so the caller skips the cloud
```
Comment (wallyworld) [inline caas/kubernetes/provider/detectcloud.go:59]: Can we check the error content to make sure it is the user group error. We probably want to still raise an error if something else goes wrong.

**Review by wallyworld — APPROVED** [review]

File: caas/kubernetes/provider/detectcloud.go (hunk 1 above)
Comment (basebandit) [reply caas/kubernetes/provider/detectcloud.go:59]: Sure thing.

Comment (basebandit) [reply caas/kubernetes/provider/detectcloud.go:61]: EDIT: The only time the error from the microk8s provider matters, is if we are bootstrapping a microk8s cloud. Else we return a notfound regardless of the error so that the caller know to skip the cloud.
